eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala

```diff
@@ -1601,15 +1760,20 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder, val
     case Event(WatchFundingSpentTriggered(tx), d: DATA_WAIT_FOR_REMOTE_PUBLISH_FUTURE_COMMITMENT) => handleRemoteSpentFuture(tx, d)
 
     case Event(WatchFundingSpentTriggered(tx), d: PersistentChannelData) =>
-      if (tx.txid == d.commitments.remoteCommit.txid) {
-        handleRemoteSpentCurrent(tx, d)
-      } else if (d.commitments.remoteNextCommitInfo.left.toOption.exists(_.nextRemoteCommit.txid == tx.txid)) {
-        handleRemoteSpentNext(tx, d)
-      } else if (tx.txid == d.commitments.localCommit.commitTxAndRemoteSig.commitTx.tx.txid) {
-        log.warning(s"processing local commit spent from the outside")
-        spendLocalCurrent(d)
+      if (d.metaCommitments.commitments.map(_.fundingTxId).contains(tx.txid)) {
+        // if the spending tx is itself a funding tx, this is a splice and there is nothing to do
```

Comment:
I think you're _mostly_ right, thanks to the fact that we always spend the latest commitment when we force-close! But there may still be a serious issue with commitments pruning:

- we have `[Commitment2, Commitment1, Commitment0]` where 1 and 2 are unconfirmed: we currently only have a `WatchFundingSpent` on `Commitment0`
- that watch triggers because `Commitment1` is in the mempool, but it's fine because we identify that it's a splice tx
- `Commitment1` and `Commitment2` then confirm in the same block
- we receive the `WatchFundingConfirmed` for `Commitment2` and prune the commitments (only `Commitment2` remains)
- since `Commitment0` was spent by `Commitment1` in that new block, the watcher will re-send a `WatchFundingSpentTriggered` for `Commitment0` 
- but we have pruned all reference to `Commitment1`, so that ends up in `handleRemoteSpentOther` which will completely mess up our state (and I believe, make us delete our published commitments based on `Commitments2`)?

This isn't trivial to reason about though, we'll need tests to exercise that and confirm our understanding of the flow.

Reply:
> we receive the WatchFundingConfirmed for Commitment2 and prune the commitments

To be pruned, commitments need to have both local/remote lock status. In your scenario the local status for `Commitment1` would be unlocked, so no commitment would be pruned.

We would prune both commitments only when we receive `WatchFundingConfirmed` for `Commitment1` (providing we also  have received remote lock).

> This isn't trivial to reason about though, we'll need tests to exercise

Definitely!

Reply:
> To be pruned, commitments need to have both local/remote lock status. In your scenario the local status for Commitment1 would be unlocked, so no commitment would be pruned.

I don't think that this is true unfortunately. First of all, our remote could very well have sent `splice_confirmed` for both commitments, because they saw the block before we did, so we only need to think about our own confirmation status.

Right now, the code in `eliminateSpliceCommitments` will delete `Commitment1` as soon as we receive the confirmation for `Commitments2`. So if we receive the `WatchFundingSpentTriggered` for `Commitment0` afterwards, I really think we're screwed, and this can happen in practice...

Reply:
I went with a cleaner version of 2. in b63243286: we simply put `WatchConfirm`s only when the previous splice tx is confirmed. This required a model/codec update, but I think it makes sense. If you like it I'll cherry-pick the change outside of this PR.

Reply:
Done in https://github.com/ACINQ/eclair/pull/2592.